### ISSUE TITLE
Check usages of `#if defined(...)`

### ISSUE THREAD

**Issue by dongcarl** [bitcoin/bitcoin#16419]
It would seem that in most cases, we want `#if ...` instead of `#if defined(...)`, so let's make sure that when we say `#if defined(...)`, we actually mean it.

Inspiration: https://github.com/bitcoin/bitcoin/pull/16344

**Comment by MarcoFalke** [comment]
`s/define/defined/`?

**Comment by Alko89** [comment]
I have gone through `bitcoin_config.h` and found a few of these.

**Comment by MarcoFalke** [comment]
Looks like https://github.com/bitcoin/bitcoin/pull/16547#issuecomment-522735491 is up for grabs?

**Comment by fanquake** [comment]
I'm going to revive #16547 and look at turning on `-Wundef`.

**Comment by luke-jr** [comment]
I looked over the uses in 8be652e43964329c1f2dadd676916b53e5c40974 and found one bug: LevelDB is using `#if defined(HAVE_O_CLOEXEC)`, which we are defining to `0` on systems without it - this leads to a build failure, and should be fixed.

**Comment by hebasto** [comment]
> I looked over the uses in [8be652e](https://github.com/bitcoin/bitcoin/commit/8be652e43964329c1f2dadd676916b53e5c40974) and found one bug: LevelDB is using `#if defined(HAVE_O_CLOEXEC)`, which we are defining to `0` on systems without it - this leads to a build failure, and should be fixed.

Fixed in #25739.

**Comment by bipulkmr-crypto** [comment]
Is this issue up for grab?

**Comment by hebasto** [comment]
> Is this issue up for grab?

It is. See #16547.

**Comment by fanquake** [comment]
> It is. See https://github.com/bitcoin/bitcoin/pull/16547.

It isn't. See #25302.

**Comment by hebasto** [comment]
On the master branch @ 4e8a7654f623fc0dad933b3d93dc54d8206c605d plus https://github.com/bitcoin/bitcoin/pull/27696, there are only two cases in our build system of explicit unconditional usage of the `AC_DEFINE` and `AC_DEFINE_UNQUOTED` macros (except for one which define string literals like `COPYRIGHT_YEAR`):

- https://github.com/bitcoin/bitcoin/blob/4e8a7654f623fc0dad933b3d93dc54d8206c605d/configure.ac#L1230
- https://github.com/bitcoin/bitcoin/blob/4e8a7654f623fc0dad933b3d93dc54d8206c605d/configure.ac#L1245

That means that the `HAVE_FDATASYNC` and `HAVE_O_CLOEXEC` are always defined, and them with `#if defined(...)`  is meaningless. All our code base uses `#if ...` for them.

 _All_ other use cases of the `AC_DEFINE` and `AC_DEFINE_UNQUOTED` macros are _conditional_ with explicitly provided value `1`. That means either a macro is defined and expands to `1` or it is not defined. Therefore, both `#if ...`  and `#if defined(...)` behave equally.

Moreover, descriptions like "Define to 1 to enable wallet functions" should be replaced with "Define this symbol to enable wallet functions" as the actual value of the defined macro does not matter at all.

---

The bottom line:
> in most cases, we want `#if ...` instead of `#if defined(...)`, so let's make sure that when we say `#if defined(...)`, we actually mean it.

These cases are `HAVE_FDATASYNC` and `HAVE_O_CLOEXEC` only, which are OK now.

---

It is worth mentioning that macros like `AC_CHECK_DECLS` define always define symbols implicitly.

---

I don't think any changes in the source files are required. However, in Developer Notes, we can recommend to use `#if defined(...)` by default except for cases when  `#if ...` is required.

**Comment by fanquake** [comment]
Going to close this now that #29876 is merged.
